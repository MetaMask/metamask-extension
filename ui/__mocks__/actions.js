const actions = require('../store/actions');
const {
  CONTRACT_ADDRESS_UNISWAP,
  CONTRACT_ADDRESS_SOURCIFY,
  CONTRACT_ADDRESS_FOUR_BYTE,
  TRANSACTION_DECODE_UNISWAP,
  TRANSACTION_DECODE_SOURCIFY,
  TRANSACTION_DECODE_FOUR_BYTE,
} = require('../../test/data/confirmations/transaction-decode');

const ERC20_TOKEN_1_MOCK = '0x2260fac5e5542a773aa44fbcfedf7c193bc2c599'; // WBTC
const ERC20_TOKEN_2_MOCK = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'; // USDC

const TOKEN_DETAILS_MOCK = {
  [ERC20_TOKEN_1_MOCK]: {
    address: ERC20_TOKEN_1_MOCK,
    standard: 'ERC20',
    decimals: 8,
    name: 'Wrapped Bitcoin',
  },
  [ERC20_TOKEN_2_MOCK]: {
    address: ERC20_TOKEN_2_MOCK,
    standard: 'ERC20',
    decimals: 6,
    name: 'USD Coin',
  },
};

module.exports = {
  ...actions,
  getTokenStandardAndDetails: (address) => {
    if (!TOKEN_DETAILS_MOCK[address]) {
      console.log('Token not found:', address);

      return Promise.resolve({
        address,
        standard: 'ERC20',
        decimals: 18,
        name: 'Missing Mock',
      });
    }

    return Promise.resolve(TOKEN_DETAILS_MOCK[address]);
  },

  // eslint-disable-next-line no-empty-function
  trackMetaMetricsEvent: () => {},

  decodeTransactionData: async (request) => {
    const { contractAddress } = request;

    if (contractAddress === CONTRACT_ADDRESS_UNISWAP) {
      return TRANSACTION_DECODE_UNISWAP;
    } else if (contractAddress === CONTRACT_ADDRESS_SOURCIFY) {
      return TRANSACTION_DECODE_SOURCIFY;
    } else if (contractAddress === CONTRACT_ADDRESS_FOUR_BYTE) {
      return TRANSACTION_DECODE_FOUR_BYTE;
    }

    return undefined;
  },
};
