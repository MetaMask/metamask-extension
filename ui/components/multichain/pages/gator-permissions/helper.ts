import { NetworkConfiguration } from '@metamask/network-controller';
import { Hex } from '@metamask/utils';

/**
 * Extracts the network name from the network configuration.
 *
 * @param networks - The network configurations.
 * @param chainId - The chain ID.
 * @param isFullNetworkName - Whether to return the full network name or not.
 * @returns The network name.
 */
export const extractNetworkName = (
  networks: Record<`0x${string}`, NetworkConfiguration>,
  chainId: Hex,
  isFullNetworkName = false,
) => {
  const network = networks[chainId];
  if (network?.name && network?.name.trim() !== '') {
    return isFullNetworkName
      ? network.name
      : `networkName${network.name.trim().split(' ')[0]}`;
  }
  return 'unknownNetworkForGatorPermissions';
};
