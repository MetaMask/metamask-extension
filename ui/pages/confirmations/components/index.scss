/** Please import your files in alphabetical order **/
@import 'advanced-gas-controls/index';
@import 'advanced-gas-fee-popover/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-gas-limit/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-defaults/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-inputs/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-inputs/base-fee-input/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-inputs/priority-fee-input/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-input-subtext/index';
@import 'approve-content-card/index';
@import 'confirmation-warning-modal/index';
@import 'confirm-page-container/index';
@import 'confirm/footer';
@import 'confirm/header/header.scss';
@import 'confirm/scroll-to-bottom';
@import 'confirm/nav/nav.scss';
@import "confirm/network-change-toast/index";
@import 'contract-details-modal/index';
@import 'custom-nonce/index';
@import 'edit-gas-display/index';
@import 'edit-gas-fee-button/index';
@import 'edit-gas-fee-popover/index';
@import 'edit-gas-fee-popover/edit-gas-item/index';
@import 'edit-gas-fee-popover/network-statistics/index';
@import 'edit-gas-fee-popover/network-statistics/status-slider/index';
@import 'edit-gas-fee-popover/edit-gas-tooltip/index';
@import 'edit-gas-popover/index';
@import 'gas-details-item/index';
@import 'multilayer-fee-message/index';
@import 'review-spending-cap/index';
@import 'set-approval-for-all-warning/index';
@import 'signature-request/index';
@import 'signature-request-original/index';
@import 'signature-request-siwe/index';
@import 'transaction-alerts/transaction-alerts';
@import 'transaction-decoding/index';
@import 'transaction-detail/index';
@import 'transaction-detail-item/index';
@import 'simulation-details/index';
