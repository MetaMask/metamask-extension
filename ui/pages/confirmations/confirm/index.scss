@use "design-system";

.confirm_wrapper {
  box-shadow: var(--shadow-size-lg) var(--color-shadow-default);

  @media screen and (min-width: design-system.$break-small) {
    max-width: 62%;
  }
}
