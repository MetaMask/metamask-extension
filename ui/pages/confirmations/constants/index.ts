import { CHAINLIST_CHAIN_IDS_MAP } from '../../../../shared/constants/network';

export const IGNORE_GAS_LIMIT_CHAIN_IDS = [CHAINLIST_CHAIN_IDS_MAP.MANTLE];

export const TYPED_SIGNATURE_VERSIONS = {
  V1: 'V1',
  V3: 'V3',
  V4: 'V4',
};

export const SPENDING_CAP_UNLIMITED_MSG = 'UNLIMITED MESSAGE';

export const TypedSignSignaturePrimaryTypes = {
  PERMIT: 'Permit',
  ORDER: 'Order',
};
