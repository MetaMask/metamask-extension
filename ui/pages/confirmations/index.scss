/** Please import your files in alphabetical order **/
@import 'confirm-approve/index';
@import 'confirm/index';
@import 'confirmation/confirmation';
@import 'send/gas-display/index';
@import 'send/send';
@import 'token-allowance/index';
@import './components/index'
