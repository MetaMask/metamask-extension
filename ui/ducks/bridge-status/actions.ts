import {
  StartPollingForBridgeTxStatusArgs,
  BridgeStatusAction,
} from '../../../shared/types/bridge-status';
import { forceUpdateMetamaskState } from '../../store/actions';
import { submitRequestToBackground } from '../../store/background-connection';
import { MetaMaskReduxDispatch } from '../../store/store';

const callBridgeStatusControllerMethod = <T extends unknown[]>(
  bridgeAction: BridgeStatusAction,
  args?: T,
) => {
  return async (dispatch: MetaMaskReduxDispatch) => {
    await submitRequestToBackground(bridgeAction, args);
    await forceUpdateMetamaskState(dispatch);
  };
};

export const startPollingForBridgeTxStatus = (
  startPollingForBridgeTxStatusArgs: StartPollingForBridgeTxStatusArgs,
) => {
  return async (dispatch: MetaMaskReduxDispatch) => {
    return dispatch(
      callBridgeStatusControllerMethod<[StartPollingForBridgeTxStatusArgs]>(
        BridgeStatusAction.START_POLLING_FOR_BRIDGE_TX_STATUS,
        [startPollingForBridgeTxStatusArgs],
      ),
    );
  };
};
