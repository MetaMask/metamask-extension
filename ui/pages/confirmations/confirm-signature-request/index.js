import PropTypes from 'prop-types';
import React, { useEffect, useState, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, withRouter } from 'react-router-dom';
import log from 'loglevel';
import { cloneDeep } from 'lodash';
import { SubjectType } from '@metamask/permission-controller';
import { TransactionStatus } from '@metamask/transaction-controller';
import * as actions from '../../../store/actions';
import txHelper from '../../../helpers/utils/tx-helper';
import SignatureRequest from '../components/signature-request';
import SignatureRequestOriginal from '../components/signature-request-original';
import SignatureRequestSIWE from '../components/signature-request-siwe';
import Loading from '../../../components/ui/loading-screen';
import { useRouting } from '../hooks/useRouting';
import {
  getTotalUnapprovedSignatureRequestCount,
  getTargetSubjectMetadata,
  getCurrentNetworkTransactions,
  getUnapprovedTransactions,
  getInternalAccounts,
  getMemoizedUnapprovedPersonalMessages,
  getMemoizedUnapprovedTypedMessages,
  getMemoizedCurrentChainId,
  getMemoizedTxId,
} from '../../../selectors';
import { useInsightSnaps } from '../../../hooks/snaps/useInsightSnaps';
import { MESSAGE_TYPE } from '../../../../shared/constants/app';
import { getSendTo } from '../../../ducks/send';

const signatureSelect = (txData, targetSubjectMetadata) => {
  const {
    type,
    msgParams: { version, siwe },
  } = txData;

  // Temporarily direct only v3 and v4 requests to new code.
  if (
    type === MESSAGE_TYPE.ETH_SIGN_TYPED_DATA &&
    (version === 'V3' || version === 'V4')
  ) {
    return SignatureRequest;
  }

  if (siwe?.isSIWEMessage && targetSubjectMetadata !== SubjectType.Snap) {
    return SignatureRequestSIWE;
  }

  return SignatureRequestOriginal;
};

const ConfirmTxScreen = ({ match }) => {
  const dispatch = useDispatch();
  const { navigateToMostRecentOverviewPage } = useRouting();
  const unapprovedMessagesTotal = useSelector(
    getTotalUnapprovedSignatureRequestCount,
  );
  const sendTo = useSelector(getSendTo);
  const internalAccounts = useSelector(getInternalAccounts);

  const { currentCurrency, blockGasLimit, signatureSecurityAlertResponses } =
    useSelector((state) => state.metamask);
  const unapprovedPersonalMsgs = useSelector(
    getMemoizedUnapprovedPersonalMessages,
  );
  const unapprovedTypedMessages = useSelector(
    getMemoizedUnapprovedTypedMessages,
  );

  const unapprovedTxs = useSelector(getUnapprovedTransactions);
  const currentNetworkTxList = useSelector(getCurrentNetworkTransactions);
  const chainId = useSelector(getMemoizedCurrentChainId);
  const index = useSelector(getMemoizedTxId);

  const [prevValue, setPrevValues] = useState();
  const history = useHistory();

  useEffect(() => {
    const unconfTxList = txHelper(
      unapprovedTxs || {},
      {},
      {},
      {},
      {},
      {},
      chainId,
    );
    if (unconfTxList.length === 0 && !sendTo && unapprovedMessagesTotal === 0) {
      navigateToMostRecentOverviewPage();
    }
  }, [
    chainId,
    navigateToMostRecentOverviewPage,
    sendTo,
    unapprovedMessagesTotal,
    unapprovedTxs,
  ]);

  useEffect(
    () => {
      if (!prevValue) {
        setPrevValues({ index, unapprovedTxs });
        return;
      }

      let prevTx;
      const { params: { id: transactionId } = {} } = match;
      if (transactionId) {
        prevTx = currentNetworkTxList.find(
          ({ id }) => `${id}` === transactionId,
        );
      } else {
        const { index: prevIndex, unapprovedTxs: prevUnapprovedTxs } =
          prevValue;
        const prevUnconfTxList = txHelper(
          prevUnapprovedTxs,
          {},
          {},
          {},
          {},
          {},
          chainId,
        );
        const prevTxData = prevUnconfTxList[prevIndex] || {};
        prevTx =
          currentNetworkTxList.find(({ id }) => id === prevTxData.id) || {};
      }

      const unconfTxList = txHelper(
        unapprovedTxs || {},
        {},
        {},
        {},
        {},
        {},
        chainId,
      );

      if (prevTx && prevTx.status === TransactionStatus.dropped) {
        dispatch(
          actions.showModal({
            name: 'TRANSACTION_CONFIRMED',
            onSubmit: () => navigateToMostRecentOverviewPage(),
          }),
        );
        return;
      }

      if (
        unconfTxList.length === 0 &&
        !sendTo &&
        unapprovedMessagesTotal === 0
      ) {
        navigateToMostRecentOverviewPage();
      }

      setPrevValues({ index, unapprovedTxs });
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );

  const { params: { id: txIdFromPath } = {} } = match;

  const txData = useMemo(() => {
    const unconfTxList = txHelper(
      unapprovedTxs || {},
      unapprovedPersonalMsgs,
      {},
      {},
      unapprovedTypedMessages,
      chainId,
    );

    log.info(`rendering a combined ${unconfTxList.length} unconf msgs & txs`);

    const unconfirmedTx = txIdFromPath
      ? unconfTxList.find(({ id }) => `${id}` === txIdFromPath)
      : unconfTxList[index];
    return unconfirmedTx ? cloneDeep(unconfirmedTx) : {};
  }, [
    chainId,
    index,
    txIdFromPath,
    unapprovedPersonalMsgs,
    unapprovedTxs,
    unapprovedTypedMessages,
  ]);

  const { warnings } = useInsightSnaps(txData.id);
  const resolvedSecurityAlertResponse =
    signatureSecurityAlertResponses?.[
      txData.securityAlertResponse?.securityAlertId
    ];

  if (resolvedSecurityAlertResponse) {
    txData.securityAlertResponse = resolvedSecurityAlertResponse;
  }

  const targetSubjectMetadata = useSelector((state) =>
    getTargetSubjectMetadata(state, txData.msgParams?.origin),
  );

  if (!txData.msgParams) {
    return <Loading />;
  }

  const SigComponent = signatureSelect(txData, targetSubjectMetadata);

  return (
    <SigComponent
      history={history}
      txData={txData}
      key={txData.id}
      accounts={internalAccounts}
      currentCurrency={currentCurrency}
      blockGasLimit={blockGasLimit}
      warnings={warnings}
    />
  );
};

ConfirmTxScreen.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
    }),
  }),
};

export default withRouter(ConfirmTxScreen);
