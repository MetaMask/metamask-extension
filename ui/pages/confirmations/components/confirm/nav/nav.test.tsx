import React from 'react';
import {
  TransactionStatus,
  TransactionType,
} from '@metamask/transaction-controller';

import { getMockConfirmState } from '../../../../../../test/data/confirmations/helper';
import { renderWithConfirmContextProvider } from '../../../../../../test/lib/confirmations/render-helpers';
import { fireEvent } from '../../../../../../test/jest';
import * as Actions from '../../../../../store/actions';
import configureStore from '../../../../../store/store';

import Nav from './nav';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => jest.fn(),
}));

const mockHistoryReplace = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    replace: mockHistoryReplace,
  }),
}));

const render = () => {
  const store = configureStore(
    getMockConfirmState({
      metamask: {
        pendingApprovals: {
          testApprovalId: {
            id: 'testApprovalId',
            time: 1528133319641,
            origin: 'metamask',
            type: 'personal_sign',
            requestData: {
              txId: 'testTransactionId',
            },
            requestState: {
              test: 'value',
            },
          },
          testApprovalId2: {
            id: 'testApprovalId2',
            time: 1528133319641,
            origin: 'metamask',
            type: 'personal_sign',
            requestData: {
              txId: 'testTransactionId',
            },
            requestState: {
              test: 'value',
            },
          },
          testApprovalId3: {
            id: 'testApprovalId3',
            time: 1528133319649,
            origin: 'metamask',
            type: 'personal_sign',
            requestData: {
              txId: 'testTransactionId',
            },
            requestState: {
              test: 'value',
            },
          },
        },
        unapprovedPersonalMsgs: {
          testApprovalId2: {
            id: 'testApprovalId2',
            type: TransactionType.personalSign,
            chainId: '0x5',
            status: TransactionStatus.unapproved,
            msgParams: {
              from: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
            },
          },
        },
      },
    }),
  );

  return renderWithConfirmContextProvider(<Nav />, store);
};

describe('ConfirmNav', () => {
  it('should match snapshot', () => {
    const { container } = render();
    expect(container).toMatchSnapshot();
  });

  it('renders the "Reject all" Button', () => {
    const { getAllByRole, getByText } = render();
    const buttons = getAllByRole('button');
    expect(buttons).toHaveLength(3);
    expect(getByText('Reject all')).toBeInTheDocument();
  });

  it('renders button to navigate to previous or next confirmation', () => {
    const { getAllByRole, getByLabelText } = render();
    const buttons = getAllByRole('button');
    expect(buttons).toHaveLength(3);
    expect(getByLabelText('Previous Confirmation')).toBeInTheDocument();
    expect(getByLabelText('Next Confirmation')).toBeInTheDocument();
  });

  it('invoke history replace method when next button is clicked', () => {
    const { getByLabelText } = render();
    const nextButton = getByLabelText('Next Confirmation');
    fireEvent.click(nextButton);
    expect(mockHistoryReplace).toHaveBeenCalledTimes(1);
  });

  it('invoke action rejectPendingApproval for all pending approvals when "Reject all" button is clicked', () => {
    const { getByRole } = render();
    const rejectAllButton = getByRole('button', { name: /Reject all/iu });
    const rejectSpy = jest
      .spyOn(Actions, 'rejectPendingApproval')
      // TODO: Replace `any` with type
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .mockImplementation(() => ({} as any));
    fireEvent.click(rejectAllButton);
    expect(rejectSpy).toHaveBeenCalledTimes(3);
  });
});
