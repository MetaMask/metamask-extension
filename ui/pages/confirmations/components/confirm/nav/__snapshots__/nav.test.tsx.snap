// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmNav should match snapshot 1`] = `
<div>
  <div
    class="mm-box mm-box--padding-3 mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-space-between mm-box--align-items-center mm-box--background-color-background-default"
    style="z-index: 2;"
  >
    <div
      class="mm-box mm-box--display-flex mm-box--align-items-center"
    >
      <button
        aria-label="Previous Confirmation"
        class="mm-box mm-button-icon mm-button-icon--size-sm mm-button-icon--disabled confirm_nav__left_btn mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-alternative mm-box--background-color-background-alternative mm-box--rounded-full"
        data-testid="confirm-nav__previous-confirmation"
        disabled=""
      >
        <span
          class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
          style="mask-image: url('./images/icons/arrow-left.svg');"
        />
      </button>
      <p
        class="mm-box mm-text mm-text--body-sm mm-box--margin-inline-2 mm-box--color-text-alternative"
      >
        1
         of 
        3
      </p>
      <button
        aria-label="Next Confirmation"
        class="mm-box mm-button-icon mm-button-icon--size-sm confirm_nav__right_btn mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-alternative mm-box--background-color-background-alternative mm-box--rounded-full"
        data-testid="confirm-nav__next-confirmation"
      >
        <span
          class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
          style="mask-image: url('./images/icons/arrow-right.svg');"
        />
      </button>
    </div>
    <button
      class="mm-box mm-text mm-button-base mm-button-base--size-md confirm_nav__reject_all mm-button-primary mm-text--body-md-medium mm-text--font-weight-normal mm-box--padding-0 mm-box--padding-right-3 mm-box--padding-left-3 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-inverse mm-box--background-color-primary-default mm-box--rounded-xl"
      data-testid="confirm-nav__reject-all"
      data-theme="light"
      type="secondary"
    >
      <span
        class="mm-box mm-icon mm-icon--size-sm mm-box--margin-inline-end-1 mm-box--display-inline-block mm-box--color-inherit"
        style="mask-image: url('./images/icons/close.svg');"
      />
      Reject all
    </button>
  </div>
</div>
`;
